Sentiment analysis can be done with or without building a machine learning model. This article will go over the Python implementation of VADERfor non-model sentiment analysis. 

After reading the article, you will learn
* What is VADER?
* How to use Python library VADER for sentiment analysis?


Let's get started!

# Step 1: Install And Import Python Libraries

The first step is to install and import Python libraries.
We need to install the `vaderSentiment` package for VADER and the`flair` package for flair.

In [2]:
# Install vaderSentiment package for VADER
!pip install vaderSentiment

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 125 kB 2.1 MB/s 


After installing the packages, let's import the Python libraries. We need to import `pandas` and `numpy` for data processing. 

For the sentiment analysis, we need to import `TextBlob`, `SentimentIntensityAnalyzer` from `vaderSentiment`, and `TextClassifier` from `flair`. We also need to load the English sentiment data from `TextClassifier` and import `Sentence` for text processing for the flair pre-trained model.

To check the sentiment prediction accuracy, we need to import `accuracy_score` from `sklearn`.

Last but not least, we set the `pandas` dataframe column width to be 1000, which will allow us to see more content from the review.

In [3]:
# Data processing
import pandas as pd
import numpy as np


# Import VADER sentiment
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

# Import accuracy_score to check performance
from sklearn.metrics import accuracy_score

# Set a wider colwith
pd.set_option('display.max_colwidth', 1000)

# Step 2: Download And Read In Data

The second step is to download and read in the dataset. 




Those who are using Google Colab for this analysis need to mount Google Drive to read the dataset. You can ignore the code below if you are not using Google Colab.

In [1]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/drive')

# Change directory
import os
os.chdir("./drive/My Drive/Colab Notebooks/3SentimentAnalysis/")

# Print out the current directory
!pwd

Mounted at /content/drive
/content/drive/My Drive/Colab Notebooks/3SentimentAnalysis


Now let's read the data into a `pandas` dataframe and see what the dataset looks like.

In [4]:
# Read in data
amz_review = pd.read_csv('amazon_cells_labelled.txt', sep='\t', names=['review', 'label'])

# Take a look at the data
amz_review.head()

review  \
0  So there is no way for me to plug it in here in the US unless I go by a converter.   
1                                                         Good case, Excellent value.   
2                                                              Great for the jawbone.   
3     Tied to charger for conversations lasting more than 45 minutes.MAJOR PROBLEMS!!   
4                                                                   The mic is great.   

   label  
0      0  
1      1  
2      1  
3      0  
4      1

`.info` helps us to get the information about the dataset. 

From the output, we can see that this data set has two columns, 1000 records, and no missing data. The 'review' column is `object` type, and the 'label'column is `int64` type.

In [5]:
# Get the dataset information
amz_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   review  1000 non-null   object
 1   label   1000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 15.8+ KB


In [6]:
amz_review.head()

review  \
0  So there is no way for me to plug it in here in the US unless I go by a converter.   
1                                                         Good case, Excellent value.   
2                                                              Great for the jawbone.   
3     Tied to charger for conversations lasting more than 45 minutes.MAJOR PROBLEMS!!   
4                                                                   The mic is great.   

   label  
0      0  
1      1  
2      1  
3      0  
4      1

Next, let's check the distribution of the label. There are 500 positive and 500 negative reviews in the dataset, so we have a balanced dataset. For a balanced dataset, we can use accuracy as the performance metric.



In [ ]:
# Check the label distribution
amz_review['label'].value_counts()

0    500
1    500
Name: label, dtype: int64

# Step 4: What is VADER?

VADER (Valence Aware Dictionary and sEntiment Reasoner) is a Python library focusing on social media sentiments. It has a built-in algorithm to change sentiment intensity based on punctuations, slang, emojis, and acronyms.

The output of VADER includes four scores: compound score, negative score, neutral score, and positive score.

* The pos, neu, and neg represent the percentage of tokens that fall into each category, so they add up together to be 100%.
* The compound score is a single score to measure the sentiment of the text. Similar to TextBlob, it ranges from -1 (extremely negative) to 1 (extremely positive). The scores near 0 represent the neural sentiment score.
* The compound score is not a simple aggregation of the pos, neu, and neg scores. Instead, it incorporates rule-based enhancements such as punctuation amplifiers.

In [7]:
# Example text
text = 'GrabNGoInfo.com is a great machine learning tutorial website.'

# VADER Sentiment
vader = SentimentIntensityAnalyzer()
vader_sentiment = vader.polarity_scores(text)
vader_sentiment

{'neg': 0.0, 'neu': 0.631, 'pos': 0.369, 'compound': 0.6249}

VADER gave the sample text 'GrabNGoInfo.com is a great machine learning tutorial website.' compound score of 0.6249. There is no negative word in the sentence, so the neg score value is 0. There are 63.1% of neutral words and 36.9% of positive words in the sentence. 

The output of VADER is saved as a dictionary. We can extract the compound sentiment score by the key 'compound'.

In [8]:
# Extract sentiment score
vader_sentiment['compound']

0.6249

Cek beberapa kalimat menggunakan vader sentiment analysis

In [10]:
text2 = 'The food is bad'
vader_sentiment2 = vader.polarity_scores(text2)
vader_sentiment2

{'neg': 0.538, 'neu': 0.462, 'pos': 0.0, 'compound': -0.5423}

In [12]:
vader_sentiment2['compound']

-0.5423

# Step 7: How To Use VADER For Sentiment Analysis

In step 7, we will apply VADER to the Amazon review dataset and see how it performs.

We first get the sentiment compound score for each review and save the values into a column called 'scores_VADER'. Then check if the compound score is positive. If the score is greater than or equal to zero, the predicted sentiment for the review is positive (labeled as 1). Otherwise, the predicted sentiment for the review is negative (labeled as 0).

In [13]:
amz_review.head()

review  \
0  So there is no way for me to plug it in here in the US unless I go by a converter.   
1                                                         Good case, Excellent value.   
2                                                              Great for the jawbone.   
3     Tied to charger for conversations lasting more than 45 minutes.MAJOR PROBLEMS!!   
4                                                                   The mic is great.   

   label  
0      0  
1      1  
2      1  
3      0  
4      1

In [14]:
# Get sentiment score for each review
vader_sentiment = SentimentIntensityAnalyzer()
amz_review['scores_VADER'] = amz_review['review'].apply(lambda s: vader_sentiment.polarity_scores(s)['compound'])

# Predict sentiment label for each review
amz_review['pred_VADER'] = amz_review['scores_VADER'].apply(lambda x: 1 if x >=0 else 0)
amz_review.head()

review  \
0  So there is no way for me to plug it in here in the US unless I go by a converter.   
1                                                         Good case, Excellent value.   
2                                                              Great for the jawbone.   
3     Tied to charger for conversations lasting more than 45 minutes.MAJOR PROBLEMS!!   
4                                                                   The mic is great.   

   label  scores_VADER  pred_VADER  
0      0       -0.3535           0  
1      1        0.8402           1  
2      1        0.6249           1  
3      0       -0.6145           0  
4      1        0.6249           1

After getting predictions from VADER, let's check the prediction accuracy. 

In [15]:
# Compare Actual and Predicted
accuracy_score(amz_review['label'],amz_review['pred_VADER'])

0.768

Comparing the actual label with the VADER prediction, we get an accuracy score of 0.768, which means that VADER predicted the review sentiment 76.8% of the time. 

TextBlob has a prediction accuracy of 68.8% for the same dataset, so VADER has an 8% improvement over the TextBlob prediction.